### ISSUE TITLE
Trouble decoding MP3 from a memory stream

### ISSUE THREAD

**Issue by Shavais** [naudio/NAudio#975]
Hello,

I hope to deploy mp3's in an encrypted zip file with my program, and stream them from the zip file at run time, so I'm trying to get streaming from a memory stream working. I have already verified that the part of my code that streams the data in from the zip file is working and the data matches the same file read from disk. I'm having trouble at the last line of this piece of code right here:

-----------
                            var frame = Mp3Frame.LoadFromStream(sourceReader.BaseStream);
                            if (frame != null)
                            {
                                if (decompressor == null)
                                {
                                    // the buffered provider doesn't know what sample rate it is working at until we have a frame
                                    WaveFormat waveFormat = new Mp3WaveFormat(
                                        frame.SampleRate,
                                        frame.ChannelMode == ChannelMode.Mono ? 1 : 2,
                                        frame.FrameLength,
                                        frame.BitRate
                                    );
                                    
                                    decompressor = new AcmMp3FrameDecompressor(waveFormat);
                                }

                                var frameBuffer = new byte[16384 * 4];
                                int decompressed = decompressor.DecompressFrame(frame, frameBuffer, 0);
-----------

In stepping through, I see that decompressed is 0, and the frameBuffer contains all zeros. It appears that the frame reading is working fine, although I notice the wafeFormat.BitsPerSample is 0. Unfortunately that field is read only. The frame says the bit rate is 414400 and the sample rate is 128000, so I guess the bits per sample are somehow less than 8? Maybe that's the problem? I wonder what's causing that..

Can you tell what I'm doing wrong?

Thanks & Regards,
Shavais

**Comment by qwbarch** [comment]
@Shavais Did you ever figure out your issue? I'm also seeing ``decompressed`` being 0 for me as well for some frames, which seems to be the reason I have hiccups when I play my audio back

**Comment by Shavais** [comment]
Nope, I gave up and switched to C# for that project. It uses NAudio and DotNetZip and it's working great.

Some years back I worked with an Intel team for a while, and one of the team members was tasked with building a multi-media viewing component for the C++ app we were working on. Several months in, he was still wrestling with bugs and problems that as I recall were not terribly dissimilar.

**Comment by markheath** [comment]
Apologies for the long delay responding. I'm trying to catch up on old issues with some AI assistance. Here's few things that might help - the good news is the fix is fairly simple:

- `waveFormat.BitsPerSample == 0` is normal for `Mp3WaveFormat` - it's a compressed-format descriptor, so there's no bits-per-sample value. That's not the problem.
- `decompressed == 0` on the first frame is also normal: an MP3 decoder has to prime on the first frame or two (bit reservoir + MDCT overlap between frames) before it produces any PCM. You keep feeding frames and output starts shortly after. @qwbarch — this is almost certainly your "hiccups" too: a 0-length decode early on (or right after a seek) is warm-up, not a dropout.

For loading an MP3 you already have fully in memory, you don't need to decode frame-by-frame at all. Decrypt the zip entry into a `MemoryStream` and pass it straight to `Mp3FileReader`, which handles framing, priming and the bit reservoir for you:

```csharp
using var ms = new MemoryStream(decryptedMp3Bytes);
using var reader = new Mp3FileReader(ms);
waveOut.Init(reader);
waveOut.Play();
```

The manual `Mp3Frame.LoadFromStream` + `AcmMp3FrameDecompressor` approach is really only for live network streaming, and even then you have to expect those priming frames that decode to 0 bytes. Glad you got it working in the end - closing this out, but the above is the recommended approach for anyone finding this later.
